README.md
# ![pybliographer-logo] Pybliographer

Pybliographer provides a framework for working with bibliographic
databases.

This software is licensed under the [GPLv2][license].

For more information, please consult the documentation at the [webiste][pybliographer-website].


## Requirements

Before installing pybliographer, please install:

* GNU Recode (v. 3.5)
* python-bibtex (also available at the [website][pybliographer-website])
* Gnome 2.x and gnome-python 2.x


## Installation 

Basically, it should be a matter of:

```
	./configure && make
	su
	make install
```

You can then run several programs :

* pybliographic : the graphical interface
* pybliographer : the text-based interface and scripting language
* pybcheck, pybconvert, pybformat, ... : several useful tools

For more configuration options, have a look at the output of `./configure --help`.


# Troubleshooting

If the configure script reports the following errors:

* *"error in python modules dependencies:  No module named ..."*
 
   This means that a required module is not installed. Unfortunately,
   depending on your distribution, the actual package to install can
   be named differently. Check on the help pages of your distribution.

* *"broken recode version"*

   Your recode package does not behave properly.  Consult
   [recode help][recode-help] for more information.


[license]: COPYING
[pybliographer-website]: https://pybliographer.org/
[recode-help]: http://pybliographer.org/RecodeHelp
[pybliographer-logo]: data/icons/48x48/pybliographic.png


NEWS
New in 1.4.0 - 2018-04-03

		* Remove dependency on python-gnome2, python-gnome2-vfs, and
          python-glade2.
		* Pybliographer now requires pygtk 2.24.0 and gettext 0.19.8
		* Improve and simplify the build system
		* Known regressions to be fixed later:
          * Format entries cannot be saved
	    * Known regressions that will not be fixed:
          * Remove support for old confirmation dialog option
          * Drop history on combo boxes

		* UI translations:
			Alan Mortensen (da)
			Anders Jonsson (sv)
			Ask Hjorth Larsen (da)
			Balázs Úr (hu)
			Daniel Mustieles (es)
			Jiri Grönroos (fi)
			Marek Cernocky (cs)
			Mario Blättermann (de)
			Piotr Dr¿g (pl)
			Rafael Fontenelle (pt_BR)
			¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿ (sr) (sr@latin)

		* Doc translations:
			Marek Cernocky (cs)
			Mario Blättermann (de)

New in 1.2.18 - 2017-10-05

       * Updated help system
       * Updated installation system
       * New icons in different sizes to follow desktop
         specifications
       * New and updated translations (Gnome I18n team)

New in 1.2.17 - 2014-11-27

       * Fixed exit issues in command line
       * Pybliographer now includes and installs an AppData file
       * Saving bibtex entries sorted
       * Minor fixes of translations, installation system
         and the GUI
       * Added Greek translation of the 
         documentation (Gnome I18n team)
       * Updated translations (Gnome I18n team)

New in 1.2.16 - 2013-07-30

       * Fix collision names with pybtex program
       * Updated translations (Gnome I18n team)

New in 1.2.15 - 2011-10-06

       * The documentation is now in Mallard format
       * Updated Medline query
       * Some Gnome cosmetics
       * Fixed DateError bug
       * New method to open documentation. Pybliographer
         now requires pygtk 2.14.0 or newer
       * Added new translations of the 
         documentation (Gnome I18n team)
       * Updated translations (Gnome I18n team)

New in 1.2.14 - 2010-10-12

       * Added German translation of the 
         documentation (Gnome I18n team)
       * Fixes on pybformat command line options
       * Documentation migrated to use gnome-doc-utils
       * Fixed path expansion bug of localedir
       * Updated Tooltips objects. Pybliographer now requires
         pygtk 2.12.0 or newer
       * Updated translations (Gnome I18n team)

New in 1.2.13 - 2010-07-15

       * Some minor gtk related fixes
       * Updated translations (Gnome I18n team)

New in 1.2.12 - 2008-11-26

       * Fixed the Ovid writer
       * Fixed a bug in latexutils
       * Fixed an escaping bug
       * Added a --disable-depchecks configure option to let packagers
         avoid failing checks when running under fakeroot (gnomevfs
         for instance would try to create $HOME/.gnome2 when imported)
       * Fix a native editing bug
       * Updated documentation
       * Updated translations (Gnome I18n team)

New in 1.2.11 - 2007-09-24

       * Fixed multiple creation of entries (Frédéric)
       * Fixed bug related to creating a new field (Frédéric)
       * Rearrangement of glade files (Frédéric)
       * Updated config loader (Frédéric)
       * Updated translations (Gnome I18n team)


New in 1.2.10 - 2007-01-18

       * Open medline queries in a new window (Frédéric)
       * Better proxy support for medline query (Frédéric)
       * Minor documentation update (Zoltán)
       * Updated translations (Gnome I18n team)


New in 1.2.9 - 2006-07-11

       * Toolbar search widget (Frédéric)
       * Improved tooltips (Zoltán)
       * Improved latex aux file parsing (Frédéric)
       * Fixed bugs related to unicode search queries and 
         bibtex keys (Frédéric)
       * Updated translations (Gnome I18n team)


New in 1.2.8 - 2006-01-20

       * setup-check.py fixed (Zoltán)
       * An alternative saving method added. It doesn't break
         links (Zoltán)
       * Fixed error reporting when the lyx pipe is not 
         available (Frédéric)
       * Fixed bug related to error_dialog_parented. A 
         compatibility module introduced (Peter)
       * Fixed document type selection for Save As (Frédéric)
       * Updated translations (Gnome I18n team)


New in 1.2.7 - 2005-10-11

       * Added shortcut keys for 'Add' and 'Edit' (Peter)
       * Added a file browser for url fields in the 
         edit window (Peter)
       * Added new features to help in the management and 
         use of Resources (Peter)
       * BibTeX preamble support (Peter), closes bug #1223989
       * A new bibtex field type (BIBTEX_VERBATIM) added for urls
         and references (Frédéric)
       * A new menu item and dialog has been added for opening 
         URLs (Zoltán), closes bug #1256166
       * File selector reworked (Frédéric)
       * Updated code to pygtk 2.4 (Frédéric)
       * Autosave feature (Zoltán)
       * Backup toggling (Zoltán)
       * Hitting escape in the main window resets the search 
         filter (Frédéric)
       * Improvements of the format dialog (Frédéric)
       * Added a small notification of the current query (Frédéric)
       * Fixes of sort and isifile issues, and config editor
         (Peter, Frédéric) 
       * Corrected confirmation request when closing a
         modified file (with compatibility switch) (Peter)
       * i18n fixes (Zoltán)
       * Updated translations (Gnome I18n team)


New in 1.2.6 - 2005-01-29

       * Fixed parsing of .aux files in pybcompact script (Peter)
       * Miscellaneous fixes in the Ovid import, closes 
         bugs # 1081186,  1078811, 1043582 (Peter)
       * All the popups can be closed with escape (Frédéric)
       * Parentheses are now allowed in BibTeX keys, closes 
         bug # #600394  (Frédéric)
       * Fixed crossref handling in pybcompact script (Frédéric)
       * Restored Search as you type (Frédéric)
       * German translation update (Frank Arnold)
       * Sort code reworked (partially, Peter)

Pyblio/Query.py
# -*- coding: utf-8 -*-
# This file is part of pybliographer
#
# Copyright (C) 2018 Germán Poo-Caamaño <gpoo@gnome.org>
# Copyright (C) 1998-2004 Frederic GOBRY <gobry@pybliographer.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#


"""
Search a keyword in a medline database

This code has been contributed by: John Vu <jvu001@umaryland.edu>

Updated by Z. Kota in 2011 for eutils. 
"""

# The time module is added for querying date ranges of publications
import os
import urllib
import urllib2
import time
import tempfile

from xml.dom.minidom import parse


query_url = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi'
fetch_url = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi'

toolName = 'pybliographer'
adminEmail = 'webmaster@pybliographer.org'


def query_info (searchterm, field, displaynum, displaystart, edate):
    if edate != 'Entrez Date':
        params = urllib.urlencode ({
            'db': 'pubmed',
            'term' : searchterm,  # searchterm is user inputted text, modified by limits if applied
            'field' : field,
            'reldate' : edate,  # two different search options given, depending on whether the user provides a relative entrez date
            'usehistory' : 'y',
            'tool' : toolName,
            'email' : adminEmail
            })
    else:
        params = urllib.urlencode ({
            'db': 'pubmed',
            'term' : searchterm,  # searchterm is user inputted text, modified by limits
            'field' : field,
            'usehistory' : 'y',
            'tool' : toolName,
            'email' : adminEmail
            })

    f = urllib2.urlopen("%s?%s" % (query_url, params))

    myquery = parse (f)
    count = myquery.getElementsByTagName('Count')[0].firstChild.data
    querykey = myquery.getElementsByTagName('QueryKey')[0].firstChild.data
    webenv = myquery.getElementsByTagName('WebEnv')[0].firstChild.data

    f.close ()
    return count, querykey, webenv


def medline_query (keyword,maxcount,displaystart,field,abstract,epubahead,pubtype,language,subset,agerange,humananimal,gender,entrezdate,pubdate,fromdate,todate):
    # note all the parameters needed to perform the query
    # Search with field limits
    if field == 'All Fields': field = 'ALL'
    elif field == 'Affiliation': field = 'AFFL'
    elif field == 'Author Name': field = 'AUTH'
    elif field == 'EC/RN Number': field = 'ECNO'
    elif field == 'Entrez Date': field = 'EDAT'
    elif field == 'Filter': field = 'FLTR'
    elif field == 'Issue': field = 'ISS'
    elif field == 'Journal Name': field = 'JOUR'
    elif field == 'Language': field = 'LANG'
    elif field == 'MeSH Date': field = 'MHDA'
    elif field == 'MeSH Major Topic': field = 'MAJR'
    elif field == 'MeSH Subheading': field = 'SUBH'
    elif field == 'MeSH Terms': field = 'MESH'
    elif field == 'Pagination': field = 'PAGE'
    elif field == 'Publication Date': field = 'PDAT'
    elif field == 'Publication Type': field = 'PTYP'
    elif field == 'Secondary Source ID': field = 'SI'
    elif field == 'Substance Name': field = 'SUBS'
    elif field == 'Text Word': field = 'WORD'
    elif field == 'Title': field = 'TITL'
    elif field == 'Title/Abstract': field = 'TIAB'
    elif field == 'UID': field = 'UID'
    elif field == 'Volume': field = 'VOL'

    # Below is added to keyword if user wants items with abstracts only
    if abstract: keyword = keyword + ' AND hasabstract'

    # Below is added to keyword if user wants items that are listed on pubmed ahead of print
    if epubahead: keyword = keyword + ' AND pubstatusaheadofprint'
    
    # Below are publication type limits to add to keyword
    if pubtype == 'Addresses': keyword = keyword + ' AND addresses[pt]'
    elif pubtype == 'Bibliography': keyword = keyword + ' AND bibliography[pt]'
    elif pubtype == 'Biography': keyword = keyword + ' AND biography[pt]'
    elif pubtype == 'Classical Article': keyword = keyword + ' AND classical article[pt]'
    elif pubtype == 'Clinical Conference': keyword = keyword + ' AND clinical conference[pt]'
    elif pubtype == 'Clinical Trial': keyword = keyword + ' AND clinical trial[pt]'
    elif pubtype == 'Clinical Trial, Phase I': keyword = keyword + ' AND clinical trial, phase I[pt]'
    elif pubtype == 'Clinical Trial, Phase II': keyword = keyword + ' AND clinical trial, phase II[pt]'
    elif pubtype == 'Clinical Trial, Phase III': keyword = keyword + ' AND clinical trial, phase III[pt]'
    elif pubtype == 'Clinical Trial, Phase IV': keyword = keyword + ' AND clinical trial, phase IV[pt]'
    elif pubtype == 'Comment': keyword = keyword + ' AND comment[pt]'
    elif pubtype == 'Congresses': keyword = keyword + ' AND congresses[pt]'
    elif pubtype == 'Consensus Development Conference': keyword = keyword + ' AND consensus development conference[pt]'
    elif pubtype == 'Consensus Development Conference, NIH': keyword = keyword + ' AND consensus development conference, NIH[pt]'
    elif pubtype == 'Controlled Clinical Trial': keyword = keyword + ' AND controlled clinical trial[pt]'
    elif pubtype == 'Corrected and Republished Article': keyword = keyword + ' AND corrected and republished article[pt]'
    elif pubtype == 'Dictionary': keyword = keyword + ' AND dictionary[pt]'
    elif pubtype == 'Directory': keyword = keyword + ' AND directory[pt]'
    elif pubtype == 'Duplicate Publication': keyword = keyword + ' AND duplicate publication[pt]'
    elif pubtype == 'Editorial': keyword = keyword + ' AND editorial[pt]'
    elif pubtype == 'Evaluation Studies': keyword = keyword + ' AND evaluation studies[pt]'
    elif pubtype == 'Festschrift': keyword = keyword + ' AND festschrift[pt]'
    elif pubtype == 'Government Publications': keyword = keyword + ' AND government publications[pt]'
    elif pubtype == 'Guideline': keyword = keyword + ' AND guideline[pt]'
    elif pubtype == 'Historical Article': keyword = keyword + ' AND historical article[pt]'
    elif pubtype == 'Interview': keyword = keyword + ' AND interview[pt]'
    elif pubtype == 'Journal Article': keyword = keyword + ' AND journal article[pt]'
    elif pubtype == 'Lectures': keyword = keyword + ' AND lectures[pt]'
    elif pubtype == 'Legal Cases': keyword = keyword + ' AND legal cases[pt]'
    elif pubtype == 'Legislation': keyword = keyword + ' AND legislation[pt]'
    elif pubtype == 'Letter': keyword = keyword + ' AND letter[pt]'
    elif pubtype == 'Meta-Analysis': keyword = keyword + ' AND meta-analysis[pt]'
    elif pubtype == 'Multicenter Study': keyword = keyword + ' AND multicenter study[pt]'
    elif pubtype == 'News': keyword = keyword + ' AND news[pt]'
    elif pubtype == 'Newspaper Article': keyword = keyword + ' AND newspaper article[pt]'
    elif pubtype == 'Overall': keyword = keyword + ' AND overall[pt]'
    elif pubtype == 'Periodical Index': keyword = keyword + ' AND periodical index[pt]'
    elif pubtype == 'Practice Guideline': keyword = keyword + ' AND practice guideline[pt'
    elif pubtype == 'Published Erratum': keyword = keyword + ' AND published erratum[pt]'
    elif pubtype == 'Randomized Controlled Trial': keyword = keyword + ' AND randomized controlled trial[pt]'
    elif pubtype == 'Retraction of Publication': keyword = keyword + ' AND retraction of publication[pt]'
    elif pubtype == 'Retracted Publication': keyword = keyword + ' AND retracted publication[pt]'
    elif pubtype == 'Review': keyword = keyword + ' AND review[pt]'
    elif pubtype == 'Review, Academic': keyword = keyword + ' AND review, academic[pt]'
    elif pubtype == 'Review Literature': keyword = keyword + ' AND review, literature[pt]'
    elif pubtype == 'Review, Multicase': keyword = keyword + ' AND review, multicase[pt]'
    elif pubtype == 'Review of Reported Cases': keyword = keyword + ' AND review of reported cases[pt]'
    elif pubtype == 'Review, Tutorial': keyword = keyword + ' AND review, tutorial[pt]'
    elif pubtype == 'Scientific Integrity Review': keyword = keyword + ' AND scientific integrity review[pt]'
    elif pubtype == 'Technical Report': keyword = keyword + ' AND technical report[pt]'
    elif pubtype == 'Twin Study': keyword = keyword + ' AND twin study[pt]'
    elif pubtype == 'Validation Studies': keyword = keyword + ' AND validation studies[pt]'
    
    # Below are language limits to add to keyword if chosen

    if language == 'English': keyword = keyword + ' AND english[la]'
    elif language == 'French': keyword = keyword + ' AND french[la]'
    elif language == 'German': keyword = keyword + ' AND german[la]'
    elif language == 'Italian': keyword = keyword + ' AND italian[la]'
    elif language == 'Japanese': keyword = keyword + ' AND japanese[la]'
    elif language == 'Russian': keyword = keyword + ' AND russian[la]'
    elif language == 'Spanish': keyword = keyword + ' AND spanish[la]'

    # Below are subset limits to add to keyword if chosen

    if subset == 'AIDS': keyword = keyword + ' AND aids[sb]'
    elif subset == 'AIDS/HIV journals': keyword = keyword + ' AND jsubsetx' #X -  AIDS/HIV journals, non-Index Medicus 
    elif subset == 'Bioethics': keyword = keyword + ' AND bioethics[ab]'
    elif subset == 'Bioethics journals': keyword = keyword + ' AND jsubsete' #E -  bioethics journals, non-Index Medicus
    elif subset == 'Biotechnology journals': keyword = keyword + ' AND jsubsetb' #B -  biotechnology journals (assigned 1990 - 1998), non-Index Medicus 
    elif subset == 'Communication disorders journals': keyword = keyword + ' AND jusbsetc' #C -  communication disorders journals (assigned 1977 - 1997), non-Index Medicus 
    elif subset == 'Complementary and Alternative Medicine': keyword = keyword + ' AND cam[sb]'
    elif subset == 'Consumer health journals': keyword = keyword + ' AND jsubsetk' #K -  consumer health journals, non-Index Medicus 
    elif subset == 'Core clinical journals': keyword = keyword + ' AND jsubsetaim' #AIM - Abridged Index Medicus A list of core clinical journals created 20 years ago 
    elif subset == 'Dental journals': keyword = keyword + ' AND jsubsetd' #D  -  dentistry journals 
    elif subset == 'Health administration journals': keyword = keyword + ' AND jsubseth' #H -  health administration journals, non-Index Medicus 
    elif subset == 'Health tech assesment journals': keyword = keyword + ' AND jsubsett'#T -  health technology assessment journals, non-Index Medicus 
    elif subset == 'History of Medicine': keyword = keyword + ' AND history[sb]'
    elif subset == 'History of Medicine journals': keyword = keyword + ' AND jsubsetq' #Q -  history of medicine journals, non-Index Medicus 
    elif subset == 'In process': keyword = keyword + ' AND in process[sb]'
    elif subset == 'Index Medicus journals': keyword = keyword + ' AND jsubsetim' #IM -  Index Medicus journals 
    elif subset == 'MEDLINE': keyword = keyword + ' AND medline[sb]'
    elif subset == 'NASA journals': keyword = keyword + ' AND jsubsets' #S -  National Aeronautics and Space Administration (NASA) journals, non-Index Medicus 
    elif subset == 'Nursing journals': keyword = keyword + ' AND jsubsetn' #N  -  nursing journals 
    elif subset == 'PubMed Central': keyword = keyword + ' AND medline pmc[sb]'
    elif subset == 'Reproduction journals': keyword = keyword + ' AND jsubsetr' #R -  reproduction journals (assigned 1972 - 1979), non-Index Medicus
    elif subset == 'Space Life Sciences': keyword = keyword + ' AND space[sb]'
    elif subset == 'Supplied by Publisher': keyword = keyword + ' AND publisher[sb]'
    elif subset == 'Toxicology': keyword = keyword + ' AND tox[sb]'

    # Age range will be added to keyword if desired
    if agerange == 'All Infant: birth-23 month': keyword = keyword + ' AND infant[mh]'
    elif agerange == 'All Child: 0-18 years': keyword = keyword + ' AND child[mh]'
    elif agerange == 'All Adult: 19+ years': keyword = keyword + ' AND adult[mh]'
    elif agerange == 'Newborn: birth-1 month': keyword = keyword + ' AND infant, newborn[mh]'
    elif agerange == 'Infant: 1-23 months': keyword = keyword + ' AND infant[mh]'
    elif agerange == 'Preschool Child: 2-5 years': keyword = keyword + ' AND child, preschool[mh]'
    elif agerange == 'Child: 6-12 years': keyword = keyword + ' AND child[mh]'
    elif agerange == 'Adolescent: 13-18 years': keyword = keyword + ' AND adolescence[mh]'
    elif agerange == 'Adult: 19-44 years': keyword = keyword + ' AND adult[mh]'
    elif agerange == 'Middle Aged: 45-64 years': keyword = keyword + ' AND middle age[mh]'
    elif agerange == 'Aged: 65+ years': keyword = keyword + ' AND aged[mh]'
    elif agerange == '80 and over: 80+ years': keyword = keyword + ' AND aged, 80 and over[mh]'

    # Human or animal studies limit will be added to keyword if desired
    if humananimal == 'Human': keyword = keyword + ' AND human[mh]'
    elif humananimal == 'Animal': keyword = keyword + ' AND animal[mh]'

    # Studies done on either females or males will be a limit of keyword
    if gender == 'Female': keyword = keyword + ' AND female[mh]'
    elif gender == 'Male': keyword = keyword + ' AND male[mh]'
    
    # Past Entrez date range will be added to keyword; the number is the relative number of days prior to today
    if entrezdate == '30 Days': entrezdate = '30'
    elif entrezdate == '60 Days': entrezdate = '60'
    elif entrezdate == '90 Days': entrezdate = '90'
    elif entrezdate == '180 Days': entrezdate = '180'
    elif entrezdate == '1 Year': entrezdate = '365'
    elif entrezdate == '2 Years': entrezdate = '730'
    elif entrezdate == '5 Years': entrezdate = '1825'
    elif entrezdate == '10 Years': entrezdate = '3650'

    # if date limits are provided, then the following will be added to keyword
    # I will only allow this if the relative entrez date is not specified above, hence the elif command
    # This is where I used the time function, gmtime() to get the current global mean time
    elif fromdate != '':
        if todate == '':
            if pubdate == 'Publication Date': keyword = keyword + ' ' + fromdate + ':' + time.strftime('%Y/%m/%d', time.gmtime()) + '[dp]'
            elif pubdate == 'Entrez Date': keyword = keyword + ' ' + fromdate +':' + time.strftime('%Y/%m/%d',time.gmtime()) + '[edat]'
        else:
            if pubdate == 'Publication Date': keyword = keyword + ' ' + fromdate + ':' + todate + '[dp]'
            elif pubdate == 'Entrez Date': keyword = keyword + ' ' + fromdate + ':' + todate + '[edat]'


    # Below is the actual call to the URL (PubMed's cgi): first to gain the count, querykey, and webenv
    # parameters, and then to get the entries that is passed to pyblio to open

    try:
        count, querykey, webenv = query_info (keyword, field, maxcount, displaystart, entrezdate) # pubmed query using the history option

    except:
        # print sys.exc_info()[:2]
        return -1


    if count == '0' : return None

 
    params = urllib.urlencode ({
        'db'       : 'pubmed',
        'rettype'  : 'medline',
        'retmode'  : 'text',
        'retmax'   : maxcount,
        'retstart' : displaystart - 1,
        'query_key': querykey,
        'WebEnv'   : webenv,
        'tool'     : toolName,
        'email'    : adminEmail
        })


    url = "%s?%s" % (fetch_url, params)

    content = urllib2.urlopen(url)
    fd, fn = tempfile.mkstemp('.medline', 'pyblio-')
    while 1:
        data = content.read(8192)
        if not data:
            break
        os.write(fd, data)
    os.close(fd)

    return fn
